Gradle 8.1-RC breaks nested map properties with non-string keys without deprecation

Previously, a `@Nested MapProperty` could have keys of any type.
It has been decided to deprecate this behavior and require keys to be strings (see #23045).
However, Gradle 8.1 appears already to assume that keys will be strings.

### Expected Behavior
Using a `@Nested MapProperty` with non-string keys should call `toString` on the keys, as before.
A deprecation warning should be emitted to notify users that this will be changed in a future release.

### Current Behavior
Gradle 8.1 assumes keys can be cast to String:
```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':reproduce'.
> class java.lang.Integer cannot be cast to class java.lang.String (java.lang.Integer and java.lang.String are in module java.base of loader 'bootstrap')
```
Since there is no warning, users are surprised that a build that worked fine in 8.0 is now broken.

### Context
Works in 8.0, broken in 8.1-RC-1 and 8.1-RC-2.

The new requirement for string keys is not a problem, but this is a breaking change that requires notice and explanation.

### Steps to Reproduce 
[gradle-repro.zip](https://github.com/gradle/gradle/files/11131765/gradle-repro.zip)

Run `./gradlew reproduce`.

### Your Environment
OpenJDK 17 on macOS

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy
@@ -787,16 +787,31 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec implements Dire
         buildFile << """
             abstract class CustomTask extends DefaultTask {
                 @Nested
-                abstract MapProperty<Integer, Object> getMap()
+                abstract MapProperty<Integer, Object> getLazyMap()
+
+                @Nested
+                Map<Integer, Object> eagerMap = [:]
+
+                @OutputFile
+                abstract RegularFileProperty getOutputFile()
+
+                @TaskAction
+                void execute() {
+                    outputFile.getAsFile().get() << lazyMap.get()
+                    outputFile.getAsFile().get() << eagerMap
+                }
             }
 
             tasks.register("customTask", CustomTask) {
-                map.put(100, "example")
+                lazyMap.put(100, "example")
+                eagerMap.put(100, "example")
+                outputFile = file("output.txt")
             }
         """
 
         expect:
         succeeds("customTask")
+        file("output.txt").text == "[100:example][100:example]"
     }
 
 

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy
@@ -782,6 +782,23 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec implements Dire
         output.contains("Input property 'nested.key1' has been removed for task ':myTask'")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/24594")
+    def "nested map with non-string key works"() {
+        buildFile << """
+            abstract class CustomTask extends DefaultTask {
+                @Nested
+                abstract MapProperty<Integer, Object> getMap()
+            }
+
+            tasks.register("customTask", CustomTask) {
+                map.put(100, "example")
+            }
+        """
+
+        expect:
+        succeeds("customTask")
+    }
+
 
     private static String namedBeanClass() {
         """

--- a/subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java
+++ b/subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java
@@ -148,13 +148,14 @@ abstract class AbstractTypeMetadataWalker<T, V extends TypeMetadataWalker.TypeMe
             );
         }
 
-        @SuppressWarnings("unchecked")
+
         @Override
         protected void walkNestedMap(Object node, String qualifiedName, BiConsumer<String, Object> handler) {
-            ((Map<String, Object>) node).forEach((key, value) -> {
+            ((Map<?, ?>) node).forEach((key, value) -> {
                 checkNotNull(key, "Null keys in nested map '%s' are not allowed.", qualifiedName);
-                checkNotNullNestedCollectionValue(qualifiedName, key, value);
-                handler.accept(key, value);
+                String stringKey = key.toString();
+                checkNotNullNestedCollectionValue(qualifiedName, stringKey, value);
+                handler.accept(stringKey, value);
             });
         }
 

